apps/web-client/src/lib/utils/misc.ts

```diff
@@ -40,3 +40,34 @@ export function normalizeName(name: string) {
 		.join(" ") // Rejoin with single spaces
 		.trim();
 }
+
+export function matchesName(needle: string, haystack: string) {
+	// Helper to normalize and tokenize the needle string into conditions
+	const getConditions = (str: string) => {
+		return str
+			.toLowerCase()
+			.replace(/[^\w\s]/g, " ") // Replace punctuation with a space
+			.replace(/\s+/g, " ") // Collapse multiple spaces into one
+			.trim()
+			.split(" ")
+			.filter((part) => part.length > 0); // Remove any empty parts resulting from normalization
+	};
+
+	const needleConditions = getConditions(needle);
+
+	// If after normalization the needle has no conditions (e.g., needle was empty or just punctuation)
+	if (needleConditions.length === 0) {
+		return needle.trim().length === 0;
+	}
+
+	// Normalize the haystack for checking (lowercase and standardize punctuation/spacing)
+	// This ensures "O'Malley" in haystack can match "OMalley" or "O Malley" from needle.
+	const normalizedHaystack = haystack
+		.toLowerCase()
+		.replace(/[^\w\s]/g, " ") // Replace punctuation with a space
+		.replace(/\s+/g, " ") // Collapse multiple spaces
+		.trim();
```

Comment:
I'm under the impression only `.toLowerCase()` is doing something useful here. What is the purpose of replacing punctuation, collapsing spaces and trimming?